### PR TITLE
Availability ui component

### PR DESCRIPTION
**Pull request by rkuykendall**
Availability component to end all availability components. First implemented in Robin, this component had the robin-specific API aspects pulled out and moved into a wrapper. This is the display-only part of the component.

After this request is merged, this component will then be imported by Robin, and then by Ostrich and Tailorbird and anyone else who needs an availability chooser. 

Next steps, after implementing in projects, will be to update the styling to match the new design mockup.

### CODE REVIEW

File: src/AvailabilityGrid/availability_grid.less [hunk 1]
```diff
@@ -0,0 +1,44 @@
+@import "~tfstyleguide/vars";
+
+.availability-grid {
+  width: 620px;
```
Comment (evanrs) [inline src/AvailabilityGrid/availability_grid.less]: @tholex Seems like we could do flexbox on this, right?

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+        })
+      })
+
+      this.setState({days: daysData});
+    }
+  },
+
+  _spitBitmap() {
+    // translate state to bitmap and return
+    let value;
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Always try to keep variables in their closure. Variables used across closures is an easy way to end up with a memory leak.

I can see that it's only used in one closure, so declare it there.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+    return this.state.days.map((dayData) => {
+      return dayData.slots.map((slot) => {
+        value = slot.selected ? '1' : '0'
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: `let value`

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    let selectionMode = 'neutral';
+
+    // precalculate the slot names one time
+    let currentSlot = moment(moment().format('YYYY-MM-DD'))
+    for (let i = 0; i < SLOTS_DAY; i++) {
+      slotNames.push(currentSlot.format('hh:mm A'));
+      currentSlot.add(MINUTES_SLOT, 'm');
+    }
+
+    for (let dayIndex in days) {
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]:
``` es6
daysData = days.map((name, index) => {
    let slots = slotNames.map((name, index) => {name, index, selected: false});

    return {index, name, slots};
});
```

Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Beautiful. This fully replaces lines 92-106, correct? It looks like it does, just want to verify.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+    const MINUTES_HOUR = 60;
+    const MINUTES_SLOT = MINUTES_HOUR / this.props.slotsHour;
+    const SLOTS_DAY = HOURS_DAY * this.props.slotsHour;
+
+    let daysData = [];
+    let slotNames = [];
+    let selectionMode = 'neutral';
+
+    // precalculate the slot names one time
+    let currentSlot = moment(moment().format('YYYY-MM-DD'))
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Try to avoid writing code with side effects, this relies on incrementing the original value making it less obvious what it will be on it's final iteration. 
```es6
slotNames = new Array(SLOTS_DAY).fill().map((slot, index) => (
    moment(moment().format('YYYY-MM-DD')).
    add(MINUTES_SLOT * (index + 1)).
    format('hh:mm A');
));
```

File: src/AvailabilityGrid/availability_grid.less (hunk 1 above)
Comment (tholex) [inline src/AvailabilityGrid/availability_grid.less]: I dunno  how this looks... if it's functioning fine with an inline-block grid, it's fine let's just keep it.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+  },
+
+  handleMouseEnter() {
+    if (this.state.mouseDown === 0) {
+      this.handleSelectionModeChanged('neutral');
+    }
+  },
+
+  handleAvailabilityPost() {
+    this.props.onAvailabilityPost(this._spitBitmap());
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: `onPost` is sufficient, it's an availability grid so we probably don't need to restate it.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+        dayBitmap.split('').map((slotValue, slotIndex) => {
+          daysData[dayIndex].slots[slotIndex].selected = slotValue === '1';
+        })
+      })
+
+      this.setState({days: daysData});
+    }
+  },
+
+  _spitBitmap() {
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: cough? sneeze? lob?

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 4 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: eslint is giving me an `Unexpected token` error for `d: false});`. Can't quite discern why.

File: src/AvailabilityGrid/availability_grid.less (hunk 1 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/availability_grid.less]: It will need style tweaks before it goes into Ostrich, so I'd leave any CSS stuff for another commit.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 2 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Makes sense. Done.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 3 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Done.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 7 above)
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: This function is totally fine, for reference I probably would've written it like this:
```es6
return this.state.days.
    map((day) => day.slots.
        map((slot) => (
            new Array(4 / this.props.slotsHour + 1).
            join(slot.selected ? 1 : 0))).
        join('')).
    join('')
```
Although I think your way is probably clearer.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+const AvailabilityGridDay = React.createClass({
+  render() {
+    let slotNodes = this.props.data.slots.map((slotData) => {
+      const {data, ...other} = this.props;
+      return (
+        <AvailabilityGridSlot
+            data={slotData}
+            dayIndex={data.index}
+            {...other} />
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: :+1:

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 9]
```diff
@@ -0,0 +1,294 @@
+'use strict';
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Not needed, this is enforced through the build process.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 5 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Let me know how fixing this code block goes.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+  onMouseUp(e) {
+    this.setState({mouseDown: this.state.mouseDown - 1});
+  },
+
+  componentWillReceiveProps(nextProps) {
+    this._digestBitmap(nextProps.bitmap);
+  },
+
+  _digestBitmap(bitmap) {
+    /**
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: In ES land documentation precedes the function.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+  },
+
+  onMouseUp(e) {
+    this.setState({mouseDown: this.state.mouseDown - 1});
+  },
+
+  componentWillReceiveProps(nextProps) {
+    this._digestBitmap(nextProps.bitmap);
+  },
+
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Docs go here, before the function definition.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 9 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Is there any case in which this does need to be done, or are the ones we have in other files just legacy?

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 8 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: @ripleyaffect

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 4 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Hm, that error is now crashing the server.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 10 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Thanks! Fixing

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 11 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Thanks! Fixing

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 7 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: > cough? sneeze? lob?

I'll switch to `getBitmap`? After all, we already have a function called that, doing that.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 4 above)
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: I've updated the comment surrounding the object literal from the slots map with parens. `=> ({name, index, selected: false})`

Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Awesome!

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 5 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Awesome! Done.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 7 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Renamed.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+          onMouseEnter={this.handleMouseEnter}
+          onMouseDown={this.handleMouseDown}
+          onMouseUp={this.handleMouseUp} >
+      </div>
+    );
+  }
+})
+
+const AvailabilityGridDay = React.createClass({
+  render() {
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Add propTypes

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 13]
```diff
@@ -0,0 +1,276 @@
+const React = require('react/addons');
+const moment = require('moment');
+
+const log = require('debug')('ui:AvailabilityGrid');
+
+const AvailabilityGridSlot = React.createClass({
+  handleMouseDown() {
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: add propTypes

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+      <div className="availability-grid-day">
+        {this.props.data.name}
+        {slotNodes.slice(this.props.minSlot, this.props.maxSlot)}
+      </div>
+    );
+  }
+})
+
+const AvailabilityGrid = React.createClass({
+  getInitialState() {
```
Comment (evanrs) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: add propTypes

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+  render() {
+    let slotNames = this.state.slotNames.map((slotName) => {
+      return (
+        <div className="availability-grid-slot-name">
+          {slotName}
+        </div>
+      );
+    })
+
+    let minSlot = 28;
```
Comment (ripleyaffect) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: These should change with a change in the granularity of the grid and have an associated comment. better, they should be passed as `minHour` and `maxHour` and use that.

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 16]
```diff
@@ -0,0 +1,276 @@
+const React = require('react/addons');
+const moment = require('moment');
+
+const log = require('debug')('ui:AvailabilityGrid');
+
+const AvailabilityGridSlot = React.createClass({
+  handleMouseDown() {
+    if (this.props.data.selected) {
+      this.props.onSelectionModeChanged('unselecting', this.props.dayIndex, this.props.data.index);
```
Comment (ripleyaffect) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: destructuring `this.props` would clean this up [alot](http://hyperboleandahalf.blogspot.com/2010/04/alot-is-better-than-you-at-everything.html).

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+      this.props.onSlotSelected(this.props.dayIndex, this.props.data.index);
+    }
+  },
+
+  handleMouseUp() {
+    this.props.onSelectionModeChanged('neutral');
+  },
+
+  handleMouseEnter() {
+    if (this.props.mouseDown === 1) {
```
Comment (ripleyaffect) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: ditto on destructuring.

Comment (ripleyaffect) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: This can also just be `if (mouseDown)` but up to you

File: src/AvailabilityGrid/AvailabilityGrid.jsx [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+      if (this.props.selectionMode === 'selecting') {
+        this.props.onSlotSelected(this.props.dayIndex, this.props.data.index);
+      } else if (this.props.selectionMode === 'unselecting') {
+        this.props.onSlotUnselected(this.props.dayIndex, this.props.data.index);
+      }
+    }
+  },
+
+  render() {
+    let cx = React.addons.classSet;
```
Comment (ripleyaffect) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: [`require('classnames')`](https://www.npmjs.com/package/classnames) instead. might have to `npm install --save` it

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 12 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Added, commenting on third.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 13 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Added, commenting on third.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 14 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Think I added them all correctly. Code is still running.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 6 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Done

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 15 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: The default maxSlot was 91, not even divisible by 4, so I don't even know how to test this since the bitmap robin is sending will be different and I don't think I can pass in 22.75. Let's talk about it.

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 16 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: True, it did!

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 18 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Done

File: src/AvailabilityGrid/AvailabilityGrid.jsx (hunk 17 above)
Comment (rkuykendall) [inline src/AvailabilityGrid/AvailabilityGrid.jsx]: Hm... probably gonna leave it. Not too messy and very explicit. But good idea.

**Comment by rkuykendall** [comment]
@evanrs @ripleyaffect Dealt with all code comments. How does it look?

**Comment by bgribble** [comment]
The code looks clean to me, I would just ask that the PR have an English description of what the work is, design goals and constraints, context, relevant tickets and other PRs, what's left to do after this PR is rolled out, etc.  For example, since there are already multiple availability components throughout the system, I'd appreciate a discussion of what's the same and what's different about this one compared to the others.

**Comment by rkuykendall** [comment]
Good point! Editing now.

**Comment by rkuykendall** [comment]
Done.

**Comment by ripleyaffect** [comment]
:+1:
